midend/optimizeSelectLabels.cpp

```diff
@@ -0,0 +1,428 @@
+
+    for (auto element : elementsOfRange) {
+        elementAlreadyAssigned = false;
+        // Check if this element is already assigned among previous labels
+        for (SelectLabel* previousElement : vecOfLabels) {
+            if ((previousElement->type == SelectLabel::RangeLabel) &&
+                        element->type == SelectLabel::RangeLabel) {
+                if (previousElement->rangeValue.first == element->rangeValue.first &&
+                    previousElement->rangeValue.second == element->rangeValue.second) {
+                    elementAlreadyAssigned = true;
+                    break;
+                }
+            } else {
+                if (previousElement->constValue == element->constValue) {
+                    elementAlreadyAssigned = true;
+                    break;
+                }
+            }
+        }
+
+        if (!elementAlreadyAssigned) {
+            vecOfLabels.push_back(element);
+            // Tree node contains new constant label
+            if (element->type == SelectLabel::ConstantLabel) {
+                IR::Constant* value = new IR::Constant(element->typeInfo,
+                                                        element->constValue, element->base);
+                IR::SelectCase* newSelectCase =
+                                            new IR::SelectCase(value, currentExpression->state);
+
+                 // Put new element in vector of all optimized select cases
+                newVecCasses.push_back(newSelectCase);
+
+            // Tree node contains new range label
+            } else {
+                // Create new select case
+                IR::Range* newLabel = getRange(element);
+                IR::SelectCase* newSelectCase =
+                                    new IR::SelectCase(newLabel, currentExpression->state);
+
+                // Put new element in vector of all optimized select cases
+                newVecCasses.push_back(newSelectCase);
+            }
+        }
+    }
+}
+
+void DoOptimizeSelectLabels::getElementsOfParsedRange(SelectLabel* root,
+                                                      std::vector<SelectLabel*> &elementsOfRange) {
+    if (!root)
+        return;
+
+    // If node is leaf node, put in vector of leafs
+    if (!root->left && !root->right) {
+        elementsOfRange.push_back(root);
+        return;
+    }
+    // If exists, go recursively through left child
+    if (root->left)
+       getElementsOfParsedRange(root->left, elementsOfRange);
+    // If exists, go recursively through right child
+    if (root->right)
+       getElementsOfParsedRange(root->right, elementsOfRange);
+}
+
+const IR::Node* DoOptimizeSelectLabels::postorder(IR::SelectExpression* selectExpression) {
+    auto vecCases = selectExpression->selectCases;
+    std::vector<SelectLabel*> vecOfLabels;
+    // Vector that will contain optimized select casses
+    IR::Vector<IR::SelectCase> newVecCasses;
+
+    for (auto currentExpression : vecCases) {
```

Comment:
You are never looking at the state component of the SelectCase. I would think that could be useful for the optimization too: for example, if you have two overlapping ranges you can optimize them differently if they have different states or the same state.